Grotrian diagram IPyWidget functionality

AyushiDaksh opened the pull request:
### :pencil: Description

_This PR is part of the [Grotrian Diagram project](https://github.com/users/AyushiDaksh/projects/1), part of Google Summer of Code 2023. Project description and related PRs/Issues can be found on the project page._

This PR aims to integrate the Plotly plot with IPyWidgets menus. In other words, this PR adds functionality to the Grotrian Diagram widget.

**Type:** :rocket: `feature`

Also, link issues affected by this pull request by using the keywords: `close`, `closes`, `closed`, `fix`, `fixes`, `fixed`, `resolve`, `resolves` or `resolved`. 


### :pushpin: Resources

Project Link: https://github.com/users/AyushiDaksh/projects/1
Idea Link: https://tardis-sn.github.io/gsoc_2023/ideas/
GSoC Project Link: https://summerofcode.withgoogle.com/programs/2023/projects/HxymUMRe


### :vertical_traffic_light: Testing

How did you test these changes?

- [ ] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)


### :ballot_box_with_check: Checklist

- [ ] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

MarkMageeAstro commented on docs/io/visualization/generating_widgets.ipynb:
```diff
@@ -179,11 +185,58 @@
+   "source": [
+    "## Energy Level Widget\n",
+    "\n",
+    "This widget lets you explore the visualize the various line interactions in a simulation in the form of an Energy Level Diagram.\n",
```
This widget lets you explore and visualise the various level populations and line interactions in a simulation in the for of an Energy Level Diagram.

MarkMageeAstro commented on docs/io/visualization/generating_widgets.ipynb:
```diff
@@ -179,11 +185,58 @@
+    "\n",
+    "This widget lets you explore the visualize the various line interactions in a simulation in the form of an Energy Level Diagram.\n",
+    "\n",
+    "You can select any ion, filter transitions by wavelength range and the supernova shell, to display an energy level diagram, where:\n",
```
You can select any ion present in the simulation and filter the transitions by wavelength or model shell.

MarkMageeAstro commented on docs/io/visualization/generating_widgets.ipynb:
```diff
@@ -179,11 +185,58 @@
+    "\n",
+    "You can select any ion, filter transitions by wavelength range and the supernova shell, to display an energy level diagram, where:\n",
+    "\n",
+    "- The horizontal platforms represent the energy level. The width of these platforms represent the level population\n",
```
The thickness of each platform represents its level population.

MarkMageeAstro commented on docs/io/visualization/generating_widgets.ipynb:
```diff
@@ -179,11 +185,58 @@
+    "You can select any ion, filter transitions by wavelength range and the supernova shell, to display an energy level diagram, where:\n",
+    "\n",
+    "- The horizontal platforms represent the energy level. The width of these platforms represent the level population\n",
+    "- The arrows represent the line interactions, where their width represent the number of packets which underwent that transition and the color represents the wavelength of the transition.\n",
```
..., where the thickness shows the number of packets that underwent that transition. The wavelength of the transition is given by the color.

MarkMageeAstro commented on docs/io/visualization/generating_widgets.ipynb:
```diff
@@ -179,11 +185,58 @@
+    "- The horizontal platforms represent the energy level. The width of these platforms represent the level population\n",
+    "- The arrows represent the line interactions, where their width represent the number of packets which underwent that transition and the color represents the wavelength of the transition.\n",
+    "\n",
+    "In addition to the above filters, you can also select the scaling of Y-axis (which represents the energy) and the max levels to display.\n"
```
...maximum number of levels...

MarkMageeAstro commented on docs/io/visualization/generating_widgets.ipynb:
```diff
@@ -179,11 +185,58 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To generate energy level widget, we will again use the simulation object `sim` and then display the widget:\n"
```
"...an energy level..."
Energy level widget or energy level diagram?

AyushiDaksh replied:
Changed it to "To generate the energy level diagram.."

MarkMageeAstro commented on docs/io/visualization/using_widgets.rst:
```diff
@@ -178,4 +178,37 @@ There are also several other options in the modebar which we have not explained
+    :alt: Demo of Energy Level Widget
+
+By selecting an ion on the widget, you can see its energy level diagram based
+on the last line interactions in the simulation.
```
"...diagram, which also shows information about the last line interactions experienced by packets in the simulation."

MarkMageeAstro commented on docs/io/visualization/using_widgets.rst:
```diff
@@ -178,4 +178,37 @@ There are also several other options in the modebar which we have not explained
+
+The y-axis of the plot represent energies and the horizontal levels represent the
+different energy levels. The width of these platform represent the level populations
+in those levels.
```
The y-axis of the plot represents energy while the horizontal lines show discrete energy levels. The thickness of each line presents the level population.

MarkMageeAstro commented on docs/io/visualization/using_widgets.rst:
```diff
@@ -178,4 +178,37 @@ There are also several other options in the modebar which we have not explained
+The arrows represent the line interaction, with the upward arrows
+represent excitation and vice versa. The width of the arrows represents the number
+of packets which underwent that interaction, while the color of the arrows represent
+the wavelength of the transition.
```
Arrows represent the line interactions experienced by packets. Upwards arrows show excitation from lower energy levels to higher levels and downward arrows show de-excitation from higher energy levels to lower levels. The thickness of each arrow represents the number of packets that underwent that interaction. The wavelength of the transition is given by the color.

MarkMageeAstro commented on docs/io/visualization/using_widgets.rst:
```diff
@@ -178,4 +178,37 @@ There are also several other options in the modebar which we have not explained
+the last line interactions and the level populations. If no shell is selected,
+then all the last line interactions are plotted and the level populations are
+averaged across shells. You can also set the max levels to show on the plot.
+
```
"You can select the range on which to filter the wavelength using the slider."
"...are averaged across all shells in the simulation."
"You can also set the maximum number of levels to show on the plot."

MarkMageeAstro reviewed (changes requested):
Haven't been able to review the code itself yet, but there are some minor changes to the documentation that should be implemented.

andrewfullard commented:
Can this be merged?

jamesgillanders commented:
@MarkMageeAstro @jamesgillanders @aoifeboyle(?) to review/test this and make a decision to merge or update

jamesgillanders commented:
@AyushiDaksh we are keen to merge this PR -- can you please rebase it?

MarkMageeAstro commented on docs/io/visualization/generating_widgets.ipynb:
```diff
@@ -5,16 +5,17 @@
    "metadata": {},
```
<p>These links don't work for me. Might be a simple rebasing issue</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/2328/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

AyushiDaksh replied:
@MarkMageeAstro This is the same link which is present in the latest `master` branch as well.
https://tardis-sn.github.io/tardis/using/visualization/using_widgets.html

MarkMageeAstro replied:
<p>Okay, can you open an issue?</p><p> </p>

sarthak-dv replied:
I've solved the issue in this PR: https://github.com/tardis-sn/tardis/pull/2533

MarkMageeAstro commented on docs/io/visualization/generating_widgets.ipynb:
```diff
@@ -5,16 +5,17 @@
    "metadata": {},
```
<p>I guess we have to use the American spelling to be consistent...</p><p>visualise --&gt; visualize</p><p> </p><p>"The horizontal lines represent the energy levels. The thickness of each line shows the relative population of that energy level, with thicker lines being more populated."</p><p>"The arrows represent the line interactions between levels, with the arrow direction giving the direction of the transition. The thickness of each arrow also shows the number of packets that underwent the transition while the wavelength is given by the color."</p><p> </p><p>"In addition, you can also select between linear- and log-scaling for the y-axis (which represents the energy of each level) and the maximum number of levels to display, beginning from the lowest energy levels."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tardis-sn/tardis/pull/2328/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

MarkMageeAstro commented on docs/io/visualization/using_widgets.rst:
```diff
@@ -178,4 +178,39 @@ There are also several other options in the modebar which we have not explained
+
+The y-axis of the plot represents energy while the horizontal lines show
+discrete energy levels. The thickness of each line represents the level
+population.
```
"...level population, with thicker lines...etc."

AyushiDaksh replied:
done

MarkMageeAstro commented on docs/io/visualization/using_widgets.rst:
```diff
@@ -178,4 +178,39 @@ There are also several other options in the modebar which we have not explained
+Arrows represent the line interactions experienced by packets. Upwards arrows
+show excitation from lower energy levels to higher levels and downward arrows
+show de-excitation from higher energy levels to lower levels. The thickness of
+each arrow represents the number of packets that underwent that interaction.
```
"...that interaction, with thicker lines...etc."

AyushiDaksh replied:
Done

MarkMageeAstro reviewed (changes requested):
Minor changes to documentation. Should be good to merge after rebasing. Is that Grotrian mockup notebook still necessary?

AyushiDaksh commented:
@MarkMageeAstro @jamesgillanders 
Sorry I somehow saw the comments now. Yes we're keen on merging this.
I'll rebase this branch and make the requested changes today

MarkMageeAstro commented:
Any luck with this @AyushiDaksh?

MarkMageeAstro commented:
@AyushiDaksh can you let us know if you're still interested in this?

AyushiDaksh commented:
@MarkMageeAstro Sorry I moved to a new city this month so the last couple of weeks were really busy.
I have started work on this now, will wrap it up this weekend

AyushiDaksh commented:
> Minor changes to documentation. Should be good to merge after rebasing. Is that Grotrian mockup notebook still necessary?
I have changed the docs according to the comments and rebased the branch.
The mockup notebook helps the user learn to use the tool quickly. I would suggest keeping it.

@MarkMageeAstro @jamesgillanders

MarkMageeAstro reviewed (changes requested):
One minor correction to documentation. Also the mockup notebook didn't show any figures for mean on ReviewNB. Can you generate it again?

Thanks for continuing on this @AyushiDaksh and your hard work getting it done!

AyushiDaksh commented:
@MarkMageeAstro 

> One minor correction to documentation.
Somehow I'm not able to see your comment. Can you point me to where you need the correction?

> Also the mockup notebook didn't show any figures for mean on ReviewNB. Can you generate it again?

Since the Grotrian Widget is a dynamic widget, it doesn't persist in the notebook when it gets saved. This is the same reason why other dynamic widgets don't show up on our documentation page (and we have to resort to GIFs instead).

MarkMageeAstro commented on docs/io/visualization/using_widgets.rst:
```diff
@@ -178,4 +178,40 @@ There are also several other options in the modebar which we have not explained
+show excitation from lower energy levels to higher levels and downward arrows
+show de-excitation from higher energy levels to lower levels. The thickness of
+each arrow represents the number of packets that underwent that interaction,
+with thicker lines representing more number of packets than the thin lines.
```
"more number of packets" -> "more packets"

AyushiDaksh replied:
done

MarkMageeAstro commented:
> @MarkMageeAstro
> 
> > One minor correction to documentation.
> 
> Somehow I'm not able to see your comment. Can you point me to where you need the correction?

Hmm, I can't find it either. Must not have been that important. I did spot one other small thing when looking for it thoguh, sorry.


> > Also the mockup notebook didn't show any figures for mean on ReviewNB. Can you generate it again?
> 
> Since the Grotrian Widget is a dynamic widget, it doesn't persist in the notebook when it gets saved. This is the same reason why other dynamic widgets don't show up on our documentation page (and we have to resort to GIFs instead).

Ah, I must have been confused because there were plots in the old version, but that was with the static version of the code

AyushiDaksh commented:
> Hmm, I can't find it either. Must not have been that important. I did spot one other small thing when looking for it thoguh, sorry.

No worries! I have fixed the thing you pointed.

> Ah, I must have been confused because there were plots in the old version, but that was with the static version of the code

Yeah we had a static plot during the initial stages of my project, which eventually became dynamic :)

Let me know if this is good to merge @MarkMageeAstro

AyushiDaksh commented:
Rebased with master and tested the widget locally again, all seems good.

Please look at https://github.com/tardis-sn/tardis/pull/2328#issuecomment-1988233291 and let me know if anything else is pending here.

Thanks!

MarkMageeAstro commented:
Looks good to me and also works on my local install. Thanks @AyushiDaksh!

AyushiDaksh commented:
Thanks @MarkMageeAstro !

@jamesgillanders Can you please review/approve so we can merge this?

It was great working on this project with TARDIS. Feel free to tag me in future issues/enhacements related to the Grotrian Diagram.

andrewfullard commented:
Just one conflict to resolve! Probably due to file renames recently.
